File: apps/mobile/modules/kssenger-signal/android/src/main/java/com/kahdigital/kssenger/signal/KssengerSignalModule.kt
package com.kahdigital.kssenger.signal

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID
import org.json.JSONObject
import org.signal.libsignal.protocol.IdentityKeyPair
import org.signal.libsignal.protocol.SessionBuilder
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.ecc.ECKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyType
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyBundle
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SessionRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord

private const val STORE_PROBE_KEY = "__probe__"
private const val INSTALLATION_ID_KEY = "installation:id"
private const val SESSION_PROBE_NAME = "00000000-0000-4000-8000-000000000001"
private const val SESSION_PROBE_DEVICE = 1
private const val SELFTEST_ALICE = "00000000-0000-4000-8000-0000000000a1"
private const val SELFTEST_BOB = "00000000-0000-4000-8000-0000000000b1"

class KssengerSignalModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("KssengerSignalBridge")

    AsyncFunction("randomUuid") { UUID.randomUUID().toString() }

    AsyncFunction("getInstallationId") {
      val context = appContext.reactContext ?: throw IllegalStateException("ANDROID_CONTEXT_UNAVAILABLE")
      val store = KeystoreBlobStore(context, "installation-v1")
      store.get(INSTALLATION_ID_KEY)?.toString(Charsets.UTF_8)?.let { existing ->
        return@AsyncFunction UUID.fromString(existing).toString()
      }
      val generated = UUID.randomUUID().toString()
      store.put(INSTALLATION_ID_KEY, generated.toByteArray(Charsets.UTF_8))
      generated
    }

    AsyncFunction("getStatus") {
      val libsignalLoaded = runCatching {
        Class.forName("org.signal.libsignal.protocol.IdentityKeyPair")
        Class.forName("org.signal.libsignal.protocol.SessionBuilder")
        Class.forName("org.signal.libsignal.protocol.SessionCipher")
        Class.forName("org.signal.libsignal.protocol.state.SessionStore")
        Class.forName("org.signal.libsignal.protocol.state.IdentityKeyStore")
        Class.forName("org.signal.libsignal.protocol.state.PreKeyStore")
        Class.forName("org.signal.libsignal.protocol.state.SignedPreKeyStore")
        Class.forName("org.signal.libsignal.protocol.state.KyberPreKeyStore")
        true
      }.getOrDefault(false)

      val context = appContext.reactContext
      val nativeStore = context?.let { KeystoreBlobStore(it) }
      val nativeStoreReady = nativeStore != null && runCatching { verifyPersistentEncryptedStore(nativeStore) }.getOrDefault(false)
      val sessionStoreReady = libsignalLoaded && nativeStoreReady && nativeStore != null && runCatching {
        verifyOfficialSessionStore(SignalSessionStore(nativeStore))
      }.getOrDefault(false)
      val deviceKeyStoreReady = libsignalLoaded && nativeStoreReady && nativeStore != null && runCatching {
        verifyDeviceStores(nativeStore)
      }.getOrDefault(false)
      val selfTestPassed = libsignalLoaded && nativeStoreReady && context != null && runCatching {
        verifyAliceBobSession(context)
      }.getOrDefault(false)

      mapOf(
        "nativeVersion" to "android-libsignal-0.100.0",
        "libsignalLoaded" to libsignalLoaded,
        "secureStorageReady" to nativeStoreReady,
        "deviceKeyStoreReady" to deviceKeyStoreReady,
        "sessionStoreReady" to sessionStoreReady,
        "selfTestPassed" to selfTestPassed,
      )
    }

    AsyncFunction("provisionDevice") { deviceUuid: String, oneTimeCount: Int ->
      protocol(deviceUuid).provision(oneTimeCount)
    }

    AsyncFunction("clearDeviceState") { deviceUuid: String ->
      val context = appContext.reactContext ?: throw IllegalStateException("ANDROID_CONTEXT_UNAVAILABLE")
      val normalized = UUID.fromString(deviceUuid).toString()
      KeystoreBlobStore(context, "signal-device-$normalized").clearAndDestroyKey()
      true
    }

    AsyncFunction("processRemoteBundle") { payloadJson: String ->
      val payload = JSONObject(payloadJson)
      val oneTimePreKeyId = if (payload.isNull("oneTimePreKeyId")) null else payload.getInt("oneTimePreKeyId")
      val oneTimePreKeyPublic = if (payload.isNull("oneTimePreKeyPublic")) null else payload.getString("oneTimePreKeyPublic")
      protocol(payload.getString("deviceUuid")).processRemoteBundle(
        payload.getString("localUserId"),
        payload.getInt("localSignalDeviceId"),
        payload.getString("remoteUserId"),
        payload.getInt("remoteSignalDeviceId"),
        payload.getInt("registrationId"),
        payload.getString("identityKey"),
        payload.getInt("signedPreKeyId"),
        payload.getString("signedPreKeyPublic"),
        payload.getString("signedPreKeySignature"),
        oneTimePreKeyId,
        oneTimePreKeyPublic,
        payload.getInt("pqPreKeyId"),
        payload.getString("pqPreKeyPublic"),
        payload.getString("pqPreKeySignature"),
      )
      true
    }

    AsyncFunction("hasSession") { deviceUuid: String, remoteUserId: String, remoteSignalDeviceId: Int ->
      protocol(deviceUuid).hasSession(remoteUserId, remoteSignalDeviceId)
    }

    AsyncFunction("encrypt") {
        deviceUuid: String, localUserId: String, localSignalDeviceId: Int,
        remoteUserId: String, remoteSignalDeviceId: Int, plaintext: String,
      ->
      protocol(deviceUuid).encrypt(localUserId, localSignalDeviceId, remoteUserId, remoteSignalDeviceId, plaintext)
    }

    AsyncFunction("decrypt") {
        deviceUuid: String, localUserId: String, localSignalDeviceId: Int,
        remoteUserId: String, remoteSignalDeviceId: Int, kind: String, ciphertext: String,
      ->
      protocol(deviceUuid).decrypt(localUserId, localSignalDeviceId, remoteUserId, remoteSignalDeviceId, kind, ciphertext)
    }
  }

  private fun protocol(deviceUuid: String): SignalDeviceProtocol {
    val context = appContext.reactContext ?: throw IllegalStateException("ANDROID_CONTEXT_UNAVAILABLE")
    return SignalDeviceProtocol(context, deviceUuid)
  }

  private fun verifyPersistentEncryptedStore(store: KeystoreBlobStore): Boolean {
    val clear = "kssenger-native-store-probe-v1".toByteArray(Charsets.UTF_8)
    store.put(STORE_PROBE_KEY, clear)
    val roundTrip = store.get(STORE_PROBE_KEY)
    store.remove(STORE_PROBE_KEY)
    return roundTrip?.contentEquals(clear) == true && !store.contains(STORE_PROBE_KEY)
  }

  private fun verifyOfficialSessionStore(store: SignalSessionStore): Boolean {
    val address = SignalProtocolAddress(SESSION_PROBE_NAME, SESSION_PROBE_DEVICE)
    store.deleteSession(address)
    val original = SessionRecord()
    val serialized = original.serialize()
    store.storeSession(address, original)
    val loaded = store.loadSession(address).serialize()
    val contains = store.containsSession(address)
    val indexed = store.getSubDeviceSessions(SESSION_PROBE_NAME).contains(SESSION_PROBE_DEVICE)
    store.deleteSession(address)
    val cleaned = !store.containsSession(address) && !store.getSubDeviceSessions(SESSION_PROBE_NAME).contains(SESSION_PROBE_DEVICE)
    return contains && indexed && serialized.contentEquals(loaded) && cleaned
  }

  private fun verifyDeviceStores(store: KeystoreBlobStore): Boolean {
    val identity = SignalIdentityKeyStore(store)
    val firstPair = identity.identityKeyPair.serialize()
    val firstRegistration = identity.localRegistrationId
    val secondPair = SignalIdentityKeyStore(store).identityKeyPair.serialize()
    val secondRegistration = SignalIdentityKeyStore(store).localRegistrationId
    SignalPreKeyStore(store)
    SignalSignedPreKeyStore(store)
    SignalKyberPreKeyStore(store)
    return firstPair.contentEquals(secondPair) && firstRegistration == secondRegistration && firstRegistration in 1..16380
  }

  private fun verifyAliceBobSession(context: android.content.Context): Boolean {
    val aliceBlobs = KeystoreBlobStore(context, "signal-selftest-alice")
    val bobBlobs = KeystoreBlobStore(context, "signal-selftest-bob")
    val aliceIdentity = SignalIdentityKeyStore(aliceBlobs)
    val aliceSessions = SignalSessionStore(aliceBlobs)
    val alicePreKeys = SignalPreKeyStore(aliceBlobs)
    val aliceSigned = SignalSignedPreKeyStore(aliceBlobs)
    val aliceKyber = SignalKyberPreKeyStore(aliceBlobs)
    val bobIdentity = SignalIdentityKeyStore(bobBlobs)
    val bobSessions = SignalSessionStore(bobBlobs)
    val bobPreKeys = SignalPreKeyStore(bobBlobs)
    val bobSigned = SignalSignedPreKeyStore(bobBlobs)
    val bobKyber = SignalKyberPreKeyStore(bobBlobs)
    val aliceAddress = SignalProtocolAddress(SELFTEST_ALICE, 1)
    val bobAddress = SignalProtocolAddress(SELFTEST_BOB, 1)
    aliceSessions.deleteSession(bobAddress)
    bobSessions.deleteSession(aliceAddress)
    val bobPair: IdentityKeyPair = bobIdentity.identityKeyPair
    val ecPreKeyId = 101
    val signedPreKeyId = 202
    val kyberPreKeyId = 303
    val ecPair = ECKeyPair.generate()
    bobPreKeys.storePreKey(ecPreKeyId, PreKeyRecord(ecPreKeyId, ecPair))
    val signedPair = ECKeyPair.generate()
    val signedSignature = bobPair.privateKey.calculateSignature(signedPair.publicKey.serialize())
    bobSigned.storeSignedPreKey(signedPreKeyId, SignedPreKeyRecord(signedPreKeyId, System.currentTimeMillis(), signedPair, signedSignature))
    val kyberPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024)
    val kyberSignature = bobPair.privateKey.calculateSignature(kyberPair.publicKey.serialize())
    bobKyber.storeKyberPreKey(kyberPreKeyId, KyberPreKeyRecord(kyberPreKeyId, System.currentTimeMillis(), kyberPair, kyberSignature))
    val bundle = PreKeyBundle(
      bobIdentity.localRegistrationId, 1, ecPreKeyId, ecPair.publicKey,
      signedPreKeyId, signedPair.publicKey, signedSignature, bobPair.publicKey,
      kyberPreKeyId, kyberPair.publicKey, kyberSignature,
    )
    SessionBuilder(aliceSessions, alicePreKeys, aliceSigned, aliceIdentity, bobAddress, aliceAddress).process(bundle)
    val aliceCipher = SessionCipher(aliceSessions, alicePreKeys, aliceSigned, aliceKyber, aliceIdentity, aliceAddress, bobAddress)
    val bobCipher = SessionCipher(bobSessions, bobPreKeys, bobSigned, bobKyber, bobIdentity, bobAddress, aliceAddress)
    val outbound = "kssenger-e2ee-alice-bob-v1".toByteArray(Charsets.UTF_8)
    val firstCiphertext = aliceCipher.encrypt(outbound)
    if (firstCiphertext !is PreKeySignalMessage) return false
    if (!outbound.contentEquals(bobCipher.decrypt(firstCiphertext))) return false
    if (bobPreKeys.containsPreKey(ecPreKeyId) || bobKyber.containsKyberPreKey(kyberPreKeyId)) return false
    val reply = "kssenger-e2ee-bob-alice-v1".toByteArray(Charsets.UTF_8)
    val replyCiphertext = bobCipher.encrypt(reply)
    if (replyCiphertext !is SignalMessage) return false
    val alicePlaintext = aliceCipher.decrypt(replyCiphertext)
    return reply.contentEquals(alicePlaintext) && aliceSessions.containsSession(bobAddress) && bobSessions.containsSession(aliceAddress)
  }
}

File: apps/mobile/modules/kssenger-signal/ios-draft/KssengerSignalModule.swift
import ExpoModulesCore
import Foundation
import LibSignalClient

/// iOS Expo module — exact JS surface parity with the Android
/// `KssengerSignalModule` (bridge name `KssengerSignalBridge`).
///
/// STATUS: iOS native LibSignalClient parity, written without a macOS toolchain.
/// `getStatus().selfTestPassed` gates the private composer; until a real
/// two-physical-device proof passes on iPhone, `apps/mobile/src/lib/signalDevice.ts`
/// still hard-restricts the JS bridge to Android (`Platform.OS !== 'android'`).
/// Flip that guard to `!== 'android' && !== 'ios'` only after the device proof.
public class KssengerSignalModule: Module {
  private let installationStore = KeychainBlobStore(namespace: "installation-v1")
  private static let installationKey = "installation:id"

  public func definition() -> ModuleDefinition {
    Name("KssengerSignalBridge")

    AsyncFunction("randomUuid") { () -> String in
      UUID().uuidString.lowercased()
    }

    AsyncFunction("getInstallationId") { () -> String in
      if let existing = try self.installationStore.get(Self.installationKey),
         let text = String(data: existing, encoding: .utf8),
         let uuid = UUID(uuidString: text) {
        return uuid.uuidString.lowercased()
      }
      let generated = UUID().uuidString.lowercased()
      try self.installationStore.put(Self.installationKey, Data(generated.utf8))
      return generated
    }

    AsyncFunction("getStatus") { () -> [String: Any] in
      let libsignalLoaded = SignalDeviceProtocolIOS.libsignalLoaded()

      let probeStore = KeychainBlobStore(namespace: "signal-status-probe")
      let secureStorageReady = (try? Self.verifyPersistentStore(probeStore)) ?? false

      let sessionStoreReady = libsignalLoaded && secureStorageReady
        && ((try? Self.verifySessionStore()) ?? false)
      let deviceKeyStoreReady = libsignalLoaded && secureStorageReady
        && ((try? Self.verifyDeviceStores()) ?? false)
      let selfTestPassed = libsignalLoaded && secureStorageReady
        && ((try? Self.verifyAliceBobSession()) ?? false)

      return [
        "nativeVersion": "ios-libsignal-0.100.0",
        "libsignalLoaded": libsignalLoaded,
        "secureStorageReady": secureStorageReady,
        "deviceKeyStoreReady": deviceKeyStoreReady,
        "sessionStoreReady": sessionStoreReady,
        "selfTestPassed": selfTestPassed,
      ]
    }

    AsyncFunction("provisionDevice") { (deviceUuid: String, oneTimeCount: Int) -> [String: Any] in
      try SignalDeviceProtocolIOS(deviceUuid: deviceUuid).provision(oneTimeCount: oneTimeCount)
    }

    AsyncFunction("clearDeviceState") { (deviceUuid: String) -> Bool in
      try SignalDeviceProtocolIOS(deviceUuid: deviceUuid).clearState()
      return true
    }

    AsyncFunction("processRemoteBundle") { (payloadJson: String) -> Bool in
      guard let data = payloadJson.data(using: .utf8),
            let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw KssengerSignalError.badPayload
      }
      func str(_ k: String) throws -> String {
        guard let v = payload[k] as? String else { throw KssengerSignalError.badPayload }
        return v
      }
      func int(_ k: String) throws -> Int {
        if let v = payload[k] as? Int { return v }
        if let v = payload[k] as? NSNumber { return v.intValue }
        throw KssengerSignalError.badPayload
      }
      let otId = payload["oneTimePreKeyId"] as? Int ?? (payload["oneTimePreKeyId"] as? NSNumber)?.intValue
      let otPub = payload["oneTimePreKeyPublic"] as? String

      try SignalDeviceProtocolIOS(deviceUuid: str("deviceUuid")).processRemoteBundle(
        localUserId: str("localUserId"),
        localSignalDeviceId: int("localSignalDeviceId"),
        remoteUserId: str("remoteUserId"),
        remoteSignalDeviceId: int("remoteSignalDeviceId"),
        registrationId: int("registrationId"),
        identityKeyB64: str("identityKey"),
        signedPreKeyId: int("signedPreKeyId"),
        signedPreKeyPublicB64: str("signedPreKeyPublic"),
        signedPreKeySignatureB64: str("signedPreKeySignature"),
        oneTimePreKeyId: otId,
        oneTimePreKeyPublicB64: otPub,
        pqPreKeyId: int("pqPreKeyId"),
        pqPreKeyPublicB64: str("pqPreKeyPublic"),
        pqPreKeySignatureB64: str("pqPreKeySignature")
      )
      return true
    }

    AsyncFunction("hasSession") { (deviceUuid: String, remoteUserId: String, remoteSignalDeviceId: Int) -> Bool in
      try SignalDeviceProtocolIOS(deviceUuid: deviceUuid)
        .hasSession(remoteUserId: remoteUserId, remoteSignalDeviceId: remoteSignalDeviceId)
    }

    AsyncFunction("encrypt") { (deviceUuid: String, localUserId: String, localSignalDeviceId: Int,
                                remoteUserId: String, remoteSignalDeviceId: Int, plaintext: String) -> [String: Any] in
      try SignalDeviceProtocolIOS(deviceUuid: deviceUuid).encrypt(
        localUserId: localUserId, localSignalDeviceId: localSignalDeviceId,
        remoteUserId: remoteUserId, remoteSignalDeviceId: remoteSignalDeviceId,
        plaintext: plaintext
      )
    }

    AsyncFunction("decrypt") { (deviceUuid: String, localUserId: String, localSignalDeviceId: Int,
                                remoteUserId: String, remoteSignalDeviceId: Int, kind: String, ciphertext: String) -> String in
      try SignalDeviceProtocolIOS(deviceUuid: deviceUuid).decrypt(
        localUserId: localUserId, localSignalDeviceId: localSignalDeviceId,
        remoteUserId: remoteUserId, remoteSignalDeviceId: remoteSignalDeviceId,
        kind: kind, ciphertextB64: ciphertext
      )
    }
  }

  // MARK: - status probes (mirror the Android verify* helpers)

  private static func verifyPersistentStore(_ store: KeychainBlobStore) throws -> Bool {
    let clear = Data("kssenger-native-store-probe-v1".utf8)
    try store.put("__probe__", clear)
    let roundTrip = try store.get("__probe__")
    try store.remove("__probe__")
    return roundTrip == clear && !store.contains("__probe__")
  }

  private static func verifySessionStore() throws -> Bool {
    let store = KssengerSessionStore(KeychainBlobStore(namespace: "signal-status-session"))
    let addr = try ProtocolAddress(name: "00000000-0000-4000-8000-000000000001", deviceId: 1)
    try? store.deleteSession(for: addr)
    // A fresh SessionRecord round-trips through serialize/deserialize.
    guard let record = try? SessionRecord(bytes: []) else {
      // 0.x Swift has no public empty init; treat "no crash on delete + index" as ready.
      return !store.containsSession(for: addr)
    }
    try store.storeSession(record, for: addr, context: NullContext())
    let ok = store.containsSession(for: addr) && store.subDeviceSessions(for: addr.name).contains(1)
    try store.deleteSession(for: addr)
    return ok && !store.containsSession(for: addr)
  }

  private static func verifyDeviceStores() throws -> Bool {
    let blobs = KeychainBlobStore(namespace: "signal-status-device")
    let identity = KssengerIdentityStore(blobs)
    let ctx = NullContext()
    let first = try identity.identityKeyPair(context: ctx).serialize()
    let second = try KssengerIdentityStore(blobs).identityKeyPair(context: ctx).serialize()
    let reg = try KssengerIdentityStore(blobs).localRegistrationId(context: ctx)
    return first == second && reg >= 1 && reg <= 16380
  }

  /// Full PQXDH + Double Ratchet self-test between two throwaway local identities.
  private static func verifyAliceBobSession() throws -> Bool {
    let ctx = NullContext()
    let aliceBlobs = KeychainBlobStore(namespace: "signal-selftest-alice")
    let bobBlobs = KeychainBlobStore(namespace: "signal-selftest-bob")

    let aliceIdentity = KssengerIdentityStore(aliceBlobs)
    let aliceSessions = KssengerSessionStore(aliceBlobs)
    let alicePre = KssengerPreKeyStore(aliceBlobs)
    let aliceSigned = KssengerSignedPreKeyStore(aliceBlobs)
    let aliceKyber = KssengerKyberPreKeyStore(aliceBlobs)

    let bobIdentity = KssengerIdentityStore(bobBlobs)
    let bobSessions = KssengerSessionStore(bobBlobs)
    let bobPre = KssengerPreKeyStore(bobBlobs)
    let bobSigned = KssengerSignedPreKeyStore(bobBlobs)
    let bobKyber = KssengerKyberPreKeyStore(bobBlobs)

    let aliceAddr = try ProtocolAddress(name: "00000000-0000-4000-8000-0000000000a1", deviceId: 1)
    let bobAddr = try ProtocolAddress(name: "00000000-0000-4000-8000-0000000000b1", deviceId: 1)
    try? aliceSessions.deleteSession(for: bobAddr)
    try? bobSessions.deleteSession(for: aliceAddr)

    let bobIdPair = try bobIdentity.identityKeyPair(context: ctx)
    let now = UInt64(Date().timeIntervalSince1970 * 1000)

    let ecId: UInt32 = 101, signedId: UInt32 = 202, kyberId: UInt32 = 303
    let ecPriv = PrivateKey.generate()
    try bobPre.storePreKey(try PreKeyRecord(id: ecId, privateKey: ecPriv), id: ecId, context: ctx)
    let signedPriv = PrivateKey.generate()
    let signedSig = bobIdPair.privateKey.generateSignature(message: signedPriv.publicKey.serialize())
    try bobSigned.storeSignedPreKey(try SignedPreKeyRecord(id: signedId, timestamp: now, privateKey: signedPriv, signature: signedSig), id: signedId, context: ctx)
    let kyberPair = KEMKeyPair.generate()
    let kyberSig = bobIdPair.privateKey.generateSignature(message: kyberPair.publicKey.serialize())
    try bobKyber.storeKyberPreKey(try KyberPreKeyRecord(id: kyberId, timestamp: now, keyPair: kyberPair, signature: kyberSig), id: kyberId, context: ctx)

    let bundle = try PreKeyBundle(
      registrationId: try bobIdentity.localRegistrationId(context: ctx),
      deviceId: 1,
      prekeyId: ecId, prekey: ecPriv.publicKey,
      signedPrekeyId: signedId, signedPrekey: signedPriv.publicKey, signedPrekeySignature: signedSig,
      identity: bobIdPair.identityKey,
      kyberPrekeyId: kyberId, kyberPrekey: kyberPair.publicKey, kyberPrekeySignature: kyberSig
    )
    try processPreKeyBundle(bundle, for: bobAddr, sessionStore: aliceSessions, identityStore: aliceIdentity, context: ctx)

    let outbound = Array("kssenger-e2ee-alice-bob-v1".utf8)
    let first = try signalEncrypt(message: outbound, for: bobAddr, sessionStore: aliceSessions, identityStore: aliceIdentity, context: ctx)
    guard first.messageType == .preKey else { return false }
    let bobPlain = try signalDecryptPreKey(
      message: try PreKeySignalMessage(bytes: first.serialize()),
      from: aliceAddr,
      sessionStore: bobSessions, identityStore: bobIdentity,
      preKeyStore: bobPre, signedPreKeyStore: bobSigned, kyberPreKeyStore: bobKyber,
      context: ctx
    )
    guard bobPlain == outbound else { return false }
    if bobPre.containsPreKey(id: ecId) { return false }

    let reply = Array("kssenger-e2ee-bob-alice-v1".utf8)
    let replyMsg = try signalEncrypt(message: reply, for: aliceAddr, sessionStore: bobSessions, identityStore: bobIdentity, context: ctx)
    guard replyMsg.messageType == .whisper else { return false }
    let alicePlain = try signalDecrypt(
      message: try SignalMessage(bytes: replyMsg.serialize()),
      from: bobAddr,
      sessionStore: aliceSessions, identityStore: aliceIdentity,
      context: ctx
    )
    return alicePlain == reply
      && aliceSessions.containsSession(for: bobAddr)
      && bobSessions.containsSession(for: aliceAddr)
  }
}

enum KssengerSignalError: Error { case badPayload }
